contracts/utils/cryptography/P256.sol

```diff
@@ -0,0 +1,297 @@
+    uint256 internal constant P = 0xFFFFFFFF00000001000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFF;
+    /// @dev N (order of G)
+    uint256 internal constant N = 0xFFFFFFFF00000000FFFFFFFFFFFFFFFFBCE6FAADA7179E84F3B9CAC2FC632551;
+    /// @dev A parameter of the weierstrass equation
+    uint256 internal constant A = 0xFFFFFFFF00000001000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFC;
+    /// @dev B parameter of the weierstrass equation
+    uint256 internal constant B = 0x5AC635D8AA3A93E7B3EBBD55769886BC651D06B0CC53B0F63BCE3C3E27D2604B;
+
+    /// @dev (P + 1) / 4. Useful to compute sqrt
+    uint256 private constant P1DIV4 = 0x3fffffffc0000000400000000000000000000000400000000000000000000000;
+
+    /// @dev N/2 for excluding higher order `s` values
+    uint256 private constant HALF_N = 0x7fffffff800000007fffffffffffffffde737d56d38bcf4279dce5617e3192a8;
+
+    /**
+     * @dev Verifies a secp256r1 signature using the RIP-7212 precompile and falls back to the Solidity implementation
+     * if the precompile is not available. This version should work on all chains, but requires the deployment of more
+     * bytecode.
+     *
+     * @param h - hashed message
+     * @param r - signature half R
+     * @param s - signature half S
+     * @param qx - public key coordinate X
+     * @param qy - public key coordinate Y
+     *
+     * IMPORTANT: This function disallows signatures where the `s` value is above `N/2` to prevent malleability.
+     * To flip the `s` value, compute `s = N - s`.
+     */
+    function verify(bytes32 h, bytes32 r, bytes32 s, bytes32 qx, bytes32 qy) internal view returns (bool) {
+        (bool valid, bool supported) = _tryVerifyNative(h, r, s, qx, qy);
+        return supported ? valid : verifySolidity(h, r, s, qx, qy);
+    }
+
+    /**
+     * @dev Same as {verify}, but it will revert if the required precompile is not available.
+     */
+    function verifyNative(bytes32 h, bytes32 r, bytes32 s, bytes32 qx, bytes32 qy) internal view returns (bool) {
+        (bool valid, bool supported) = _tryVerifyNative(h, r, s, qx, qy);
+        if (supported) {
+            return valid;
+        } else {
+            revert Errors.MissingPrecompile(address(0x100));
+        }
+    }
+
+    /**
+     * @dev Same as {verify}, but it will return false if the required precompile is not available.
+     */
+    function _tryVerifyNative(
+        bytes32 h,
+        bytes32 r,
+        bytes32 s,
+        bytes32 qx,
+        bytes32 qy
+    ) private view returns (bool valid, bool supported) {
+        if (r == 0 || uint256(r) >= N || s == 0 || uint256(s) > HALF_N || !isOnCurve(qx, qy)) {
+            return (false, true); // signature is invalid, and its not because the precompile is missing
+        }
+
+        (bool success, bytes memory returndata) = address(0x100).staticcall(abi.encode(h, r, s, qx, qy));
+        return (success && returndata.length == 0x20) ? (abi.decode(returndata, (bool)), true) : (false, false);
```

Comment:
What if Ethereum puts another precompile at `0x100` in the future that also returns a `0x20` length return data? Might be an edge case, but EIP and RIP must not align in the long run! Or another example is, if an L2 already put a precompile on `0x100` which is different but also returns `0x20` length (haven't done research on this tbh).

Reply:
We were told that EIP-7212 is not up to date, that RIP-7212 is the one to follow, and that precompile address allocation would be organised at the RIP level to avoid conflicts.

If the community can't figure that out, wtf are we (library devs) expected to do ?

Reply:
"would be organised at the RIP level to avoid conflicts." -> talk is cheap ;) so idk, I just don't trust this really. The reason why I haven't provided a native way to call the `P256` precompile in my snekmate (Vyper) library is that I leave this decision to the devs deploying it. I.e. they plan to deploy on 5 chains, 3 of them have the precompile and 2 not, so they write their own wrapper around it using e.g. `0x100` as target since it's ensured that it exists there.

Reply:
If you don't want to risk it, just use `verifySolidity`

Reply:
you know how people are: "hey let's gas optimise everything and native looks sweet in our tests since we run it on a `P256`-supported fork." I think a good compromise is to put a warning in the docs about the current situation?

Reply:
Please feel free to propose some change with this additional doc :)

Reply:
Are there any concrete past examples or discussion about L2s deploying Ethereum precompiles on different addresses than L1?

Reply:
Not that I'm personally aware of but there are examples where L1 precompiles are not deployed on L2. One example is ZKsync Era: https://docs.zksync.io/build/developer-reference/ethereum-differences/pre-compiles. All I'm trying to say is that we need to be prepared for more EVM fragmentation than any of us as library authors would like. Also, the past doesn't really imply the future, so not sure if this is a good indicator...

Reply:
> L1 precompiles are not deployed on L2

Yes this definitely has to be considered, that's why the default interface in this library implements a fallback to a Solidity implementation.



> I'm trying to say is that we need to be prepared for more EVM fragmentation than any of us as library authors would like.

I agree with this, but not with the idea that we should stop providing features.

Reply:
Well, actually this library is designed to work on every reasonable chain, because of the Solidity fallback. My point was that a wrapper for the precompile is valuable.

Reply:
> but not with the idea that we should not provide features unless they work on every chain.

I never said this. My main point here is that the precompile check is _not_ future proof as the return data of length 32 bytes could also come from other precompiles in the future on that address. I'm not saying this will happen, but I have seen too much shit over the last years to be paranoid enough about this edge case.

Reply:
Thanks for pointing out @pcaversaccio. I agree with the concern and now I think more about it I also agree we should explain the situation at the very minimum. We raised [a similar concern for Math.modexp](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3298#discussion_r1475446317) that ended up in [an explicit documentation mention](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/4032b42694ff6599b17ffde65b2b64d7fc8a38f8/contracts/utils/math/Math.sol#L317), so I would consider adding a note here.

Regardless, both RIP-7212 and EIP-7212 agree that the precompile "returns 1 in 32 bytes format". The current implementation reverts if the returndata can't be decoded as a boolean. In my opinion, it's unlikely that a precompile at `address(0x100)` will randomly return `true` but I recognize the possibility. Would you agree the decoding constrains the possibility?

In regards to why someone would use

Reply:
> Would you agree the decoding constrains the possibility?

Overall yes, but I want to raise that this might result in an unexpected behaviour since you don't expect a revert from a precompile usually (but no return data to indicate an error).

Reply:
I agree it's unexpected behavior, but it's a manageable risk. 

>  I think a good compromise is to put a warning in the docs about the current situation?

I hear you with this and think an "IMPORTANT" block should be enough

Reply:
sgtm!